src/DataLogger/DataLogger.cpp

```diff
@@ -0,0 +1,248 @@
+#include "settings.h"
+#include "DataLogger.h"
+#include <SD.h>
+#include <iostream>
+using namespace std;
+
+//public methods:
+
+DataLogger::DataLogger(int pinNumber, bool O_SYNC_FLAG /*= false*/)
+{
+    if (SD.begin(pinNumber, SPI1))
+    {
+        DEBUG_SERIAL_LN("SD card is present & ready");
+    }
+    else
+    {
+        DEBUG_SERIAL_LN("SD card missing or failed");
+    }
+
+    _O_SYNC = O_SYNC_FLAG;
+}
+
+DataLogger::~DataLogger(){};
+
+bool DataLogger::create(String name, int numColumns)
+{
+    File curFile = SD.open(name, FILE_WRITE);
+    _curFilePtr = &curFile;
+    _numColumns = numColumns;
+    _curColumn = 0;
+    _buffer = "";
+    
+    if (_curFilePtr != nullptr)
+    {
+        DEBUG_SERIAL_LN("Creating file successful");
+        return true;
+    }
+    else
+    {
+        DEBUG_SERIAL_LN("Creating file failed.");
+        return false;
+    }
+}
+
+bool DataLogger::open(String name, int numColumns)
+{
+    _curColumn = 0;
+    _buffer = "";
+
+    // check if file exists
+    if (SD.exists(name))
+    {
+        // file already exists
+        // open the file and see how many columns are in the csv file
+
+        DEBUG_SERIAL_LN("Loading " + name);
+
+        File tempFile;
+        tempFile = SD.open(name);
+
+        // some variables to help identify how many columns in the first line
+        int commaNum = 0;
+        bool charactersPresent = false;
+        bool escaped = false;
+
+        int columnsInFile = 0;
+
+        while (tempFile.available())
+        {
+            char readChar = tempFile.read(); // char to hold a character that has been read
+
+            if (readChar == ',')
+            {
+                commaNum++;
+            }
+            else if (readChar == '\n' || readChar == '\r')
```

Comment:
You could just check for ```'\n'``` here, which would catch both ```LF``` and ```CRLF```

Reply:
I do agree, but, this made me think..

If there was a file that was only a CRLF
```
\r\n
```
then only checking for ```'\n'``` would cause the file to be interpreted as having one column, even though it arguably would be a csv file with zero columns. Because of this, in my opinion, including the ```'\r'``` would make the code more thorough.

Reply:
That's fine.  In general I would trust that the csv file is formatted as defined [here](https://www.rfc-editor.org/rfc/rfc4180), so I wouldn't worry too much about edge cases like that.